# Khám phá dữ liệu và Tiền xử lý dữ liệu

## Các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import re

## Khám phá dữ liệu

In [6]:
manga_df = pd.read_csv('manga.csv')
manga_df.head()

manga_id                                            title   type  score  \
0         2                                          Berserk  manga   9.45   
1        13                                        One Piece  manga   9.20   
2        25                              Fullmetal Alchemist  manga   9.05   
3      1706  JoJo no Kimyou na Bouken Part 7: Steel Ball Run  manga   9.27   
4      4632                                   Oyasumi Punpun  manga   9.03   

   scored_by                status  volumes  chapters  start_date    end_date  \
0     268737  currently_publishing      NaN       NaN  1989-08-25         NaN   
1     305917  currently_publishing      NaN       NaN  1997-07-22         NaN   
2     143879              finished     27.0     116.0  2001-07-12  2010-09-11   
3     125482              finished     24.0      96.0  2004-01-19  2011-04-19   
4     141146              finished     13.0     147.0  2007-03-15  2013-11-02   

   ...                                            authors  \
0  ...  [{'id': 1868, 'first_name': 'Kentarou', 'last_...   
1  ...  [{'id': 1881, 'first_name': 'Eiichiro', 'last_...   
2  ...  [{'id': 1874, 'first_name': 'Hiromu', 'last_na...   
3  ...  [{'id': 2619, 'first_name': 'Hirohiko', 'last_...   
4  ...  [{'id': 2836, 'first_name': 'Inio', 'last_name...   

              serializations  \
0           ['Young Animal']   
1  ['Shounen Jump (Weekly)']   
2         ['Shounen Gangan']   
3             ['Ultra Jump']   
4      ['Big Comic Spirits']   

                                            synopsis  \
0  Guts, a former mercenary now known as the "Bla...   
1  Gol D. Roger, a man referred to as the "Pirate...   
2  Alchemists are knowledgeable and naturally tal...   
3  In the American Old West, the world's greatest...   
4  Punpun Onodera is a normal 11-year-old boy liv...   

                                          background  \
0  Berserk won the Award for Excellence at the si...   
1  One Piece is the highest selling manga series ...   
2  Hagane no Renkinjutsushi won the 49th Shogakuk...   
3  JoJo no Kimyou na Bouken Part 7: Steel Ball Ru...   
4  Oyasumi Punpun was originally serialized in Yo...   

                                        main_picture  \
0  https://cdn.myanimelist.net/images/manga/1/157...   
1  https://cdn.myanimelist.net/images/manga/2/253...   
2  https://cdn.myanimelist.net/images/manga/3/243...   
3  https://cdn.myanimelist.net/images/manga/3/179...   
4  https://cdn.myanimelist.net/images/manga/3/164...   

                                                 url        title_english  \
0            https://myanimelist.net/manga/2/Berserk              Berserk   
1         https://myanimelist.net/manga/13/One_Piece            One Piece   
2  https://myanimelist.net/manga/25/Fullmetal_Alc...  Fullmetal Alchemist   
3  https://myanimelist.net/manga/1706/JoJo_no_Kim...                  NaN   
4  https://myanimelist.net/manga/4632/Oyasumi_Punpun     Goodnight Punpun   

                    title_japanese  \
0                            ベルセルク   
1                        ONE PIECE   
2                           鋼の錬金術師   
3  ジョジョの奇妙な冒険 Part7 STEEL BALL RUN   
4                         おやすみプンプン   

                                      title_synonyms jikan  
0                         ['Berserk: The Prototype']  True  
1                                                 []  True  
2  ['Full Metal Alchemist', 'Hagane no Renkinjuts...  True  
3  ["JoJo's Bizarre Adventure Part 7: Steel Ball ...  True  
4                                                 []  True  

[5 rows x 31 columns]

#### Dữ liệu có bao nhiêu dòng và cột ?

In [4]:
manga_df.shape

(67273, 31)

Dữ liệu có 67273 dòng và 31 cột

#### Ý nghĩa của từng dòng là gì ?

Mỗi dòng là thông tin của một manga

#### Có tồn tại dòng nào bị trùng lắp không ?

In [5]:
np.sum(manga_df.duplicated())

0

Dữ liệu không có dòng nào trùng lắp

#### Ý nghĩa của từng cột là gì ?

  | Cột               | Ý nghĩa                                                                       |
  |-------------------|-------------------------------------------------------------------------------|
  |`manga_id`         | ID của manga (mỗi manga có một ID riêng biệt)                                 |
  |`title`            | Tựa đề của manga (viết theo phiên âm tiếng Nhật hoặc tiếng Anh)               |
  |`type`             | Được phát hành theo dạng nào (truyện tranh, tiểu thuyết,...)                  |
  |`score`            | Điểm số trên trang MyAnimeList (MAL)                                          |
  |`scored_by`        | Số lượng người dùng đánh giá manga                                            |
  |`status`           | Tình trạng (đang phát hành, đã hoàn thành, tạm dừng,...)                      |
  |`volumes`          | Số tập                                                                        |
  |`chapters`         | Số chương                                                                     |
  |`start_date`       | Ngày bắt đầu phát hành                                                        |
  |`end_date`         | Ngày kết thúc phát hành                                                       |
  |`members`          | Số lượng người dùng có manga này trong list của họ                            |
  |`favorites`        | Số lượng người dùng yêu thích manga này                                       |
  |`sfw`              | Manga có nhãn 18+ hay không                                                   |
  |`approved`         | Manga có được phê duyệt hay chưa                                              |
  |`created_at_before`| Ngày trước khi mục của manga này được tạo                                     |
  |`updated_at`       | Ngày cập nhật gần đây nhất của manga này                                      |
  |`real_start_date`  | Ngày bắt đầu phát hành thật sự                                                |
  |`real_end_date`    | Ngày kết thúc phát hành thật sự                                               |
  |`genres`           | Thể loại                                                                      |
  |`themes`           | Chủ đề                                                                        |
  |`demographics`     | Nhân khẩu học                                                                 |
  |`authors`          | Tên tác giả                                                                   |
  |`serializations`   | Danh sách các tạp chí nơi manga này được đăng lên                             |
  |`sypnosis`         | Mô tả                                                                         |
  |`background`       | Thông tin thêm                                                                |
  |`main_picture`     | URL hình ảnh chính của manga trên MAL                                         |
  |`url`              | URL của manga trên MAL                                                        |
  |`title_english`    | Tựa đề tiếng Anh của manga                                                    |
  |`title_japanese`   | Tựa đề tiếng Nhật của manga                                                   |
  |`title_synonyms`   | Danh sách các tựa đề đồng nghĩa                                               |
  |`jikan`            | Manga này được scrape từ Jikan API hay API chính thức của MAL                 |

#### Kiểu dữ liệu hiện tại của từng cột là gì ? Có tồn tại những cột nào có kiểu dữ liệu không phù hợp hay không ?

In [8]:
manga_df.dtypes

manga_id               int64
title                 object
type                  object
score                float64
scored_by              int64
status                object
volumes              float64
chapters             float64
start_date            object
end_date              object
members                int64
favorites              int64
sfw                     bool
approved                bool
created_at_before     object
updated_at            object
real_start_date       object
real_end_date         object
genres                object
themes                object
demographics          object
authors               object
serializations        object
synopsis              object
background            object
main_picture          object
url                   object
title_english         object
title_japanese        object
title_synonyms        object
jikan                   bool
dtype: object

- Cột volumes và chapters nên có kiểu dữ liệu số nguyên
- Các cột thông tin về ngày tháng (start_date, end_date, created_at_before, updated_at, real_start_date, real_end_date) nên có kiểu dữ liệu thời gian